compiler/mx.compiler/mx_graal_benchmark.py

```diff
@@ -784,6 +784,163 @@ def benchSuiteName(self):
 
 mx_benchmark.add_bm_suite(DaCapoMoveProfilingBenchmarkSuite())
 
+class DaCapoD3SBenchmarkSuite(DaCapoBenchmarkSuite): # pylint: disable=too-many-ancestors
+    """DaCapo 9.12 Bach benchmark suite implementation with D3S modifications."""
+
+    def name(self):
+        return "dacapo-d3s"
+
+    def daCapoSuiteTitle(self):
+        return "DaCapo 9.12-D3S-20180206"
+
+    def daCapoClasspathEnvVarName(self):
+        return "DACAPO_D3S_CP"
+
+    def daCapoLibraryName(self):
+        return "DACAPO_D3S"
+
+    def successPatterns(self):
+        return ['']
+
+    def resultFilter(self, values, iteration, endOfWarmupIndex):
+        """Count iterations, convert iteration time to milliseconds."""
+        # Called from lambda, increment call counter
+        iteration['value'] = iteration['value'] + 1
+        # Skip warm-up?
+        if iteration['value'] < endOfWarmupIndex:
+            return None
+
+        values['iteration_time_ms'] = str(int(values['iteration_time_ns']) / 1000 / 1000)
+        return values
+
+    def rules(self, out, benchmarks, bmSuiteArgs):
+        runArgs = self.postprocessRunArgs(benchmarks[0], self.runArgs(bmSuiteArgs))
+        if runArgs is None:
+            return []
+        totalIterations = int(runArgs[runArgs.index("-n") + 1])
+        out = [
+          mx_benchmark.CSVFixedFileRule(
+            self.resultCsvFile,
+            None,
+            {
+              "benchmark": ("<benchmark>", str),
+              "bench-suite": self.benchSuiteName(),
+              "vm": "jvmci",
+              "config.name": "default",
+              "config.vm-flags": self.shorten_vm_flags(self.vmArgs(bmSuiteArgs)),
+              "metric.name": "warmup",
+              "metric.value": ("<iteration_time_ms>", int),
+              "metric.unit": "ms",
+              "metric.type": "numeric",
+              "metric.score-function": "id",
+              "metric.better": "lower",
+              "metric.iteration": ("$iteration", int)
+            },
+            filter_fn=lambda x, counter={'value': 0}: self.resultFilter(x, counter, 0)
```

Comment:
If I understood correctly, the default parameter is initialized only once per lambda, so this is effectively creating some state associated with that lambda. My python-fu might be lacking, but are you sure this is a standard idiom? If not, maybe a better idea is to initialize counter as a local variable inside `def rules`.

Reply:
One other thing I'm slightly worried about here is that `filter_fn` might be called with the assumption that it is pure function. Will that always be the case?

https://github.com/graalvm/mx/blob/master/mx_benchmark.py#L525

Reply:
So the CSV file does not have any information about whether the row is a part of the warmup or the final measurement? In that case, let's keep it as-is.

Reply:
Exactly, the rows are not numbered or marked in any way. Perhaps that is something we should eventually add.

Reply:
Before we merge, could you please add a comment above this lambda, explaining that assumes that it's invoked in only one batch of calls, because it has internal state?